docs/designs/restore-and-gc.md

```diff
@@ -0,0 +1,160 @@
+
+Docfx uses `.lock.json`. Whenever you run docfx `restore`, docfx generates or updates your `dependency lock`, which will look something like this:
+
+```json
+{
+    "git":{
+        "https://github.com/docfx-code-sample#master":{
+            "version": "fd6968dec9a9a39aec1845232466fe35fca520da"
+        },
+        "https://github.com/docfx-code-sample#live":{
+            "version": "a82cb2a51f6925b2fa87275218b01678c5aadfe1",
+            "git":{
+                "https://github.com/azure-code-sample#live":{
+                    "version": "c9a9a39aec184523246678c5aadfea82cb2"
+                }
+            },
+            "downloads":{
+                "https:///some-blob-service/build-history.json":{
+                    "version": "66458170ba048c7fc2de4+%18b265bc4e9a6b4fdff8b5210caf9e357%22"
+                }
+            }
+        }
+    },
+    "downloads":{
+        "https:///some-blob-service/github-user-cache.json":{
+            "version": "18b265bc4e9a6bf4cb2d+%22f02fde5b4f3817266458170ba048c7fc2de46287%22"
+        }
+    }
+}
+```
+
+- `git` is the version mappings of dependency repositories
+- `downloads` is the version mappings of dependency files
+- `version` is `commit id` of the dependency repository  or `content sha1` of the dependency file
+
+This file describes an `exact`, and more importantly `reproducible` `dependencies` trees. Once it's present, any future `restore` will base its work off this file, instead of recalculating`.
+
+The presence of the `dependency lock` changes the `restore` behavior:
+
+- The dependency files/repository described by this `dependency lock` is reproduced, which means reproducing the structure/version described in the file.
+- Any missing version or dependencies are restored in the usual fashion.
+
+> Note: currently we only support `git` in `dependency lock` which means the `downloads` will be restored to latest version always, limited by the service provides these downloads.
+
+## Using locked dependencies
+
+You can setup your `.lock.json` file(local file or a URL) in the configuration like
+
+```yml
+dependencyLock: https://some-blob-service/repo/branch/.lock.json
+``` 
+
+or 
+
+```yml
+dependencyLock: .lock.json
+```
+
+Whenever you run docfx `restore`, docfx will:
+
+- [GET] **Try** to get the `.lock.json` file from configuration, if it's present, any future dependencies restore will base its work off this file.
+- [UPDATE] Once finished the restore, docfx will **try** to create/update the `.lock.json` file defined in configuration, no matter it's a local or remote location.
+
+For the last step [UPDATE], you can use command `--no-save` to skip.
```

Comment:
`npm` generates lock files by default, so it has a `--no-save` shortcut to disable it. We are quite the opposite: we don't generate lock files by default, unless you set `dependencyLock` somewhere, so maybe we don't need this option.